[WIP] improvement suggestions for the label guide documentation (until sorting dimensions chapter) (#1699)
arviz-devs/arviz: doc/source/user_guide/label_guide.rst

Label guide
===========

Basic labelling
---------------

All ArviZ plotting functions and some stats functions can take an optional ``labeller`` argument.
By default, labels show the variable name.
Multidimensional variables also show the coordinate value.

.. ipython::

Example: Default labelling
~~~~~~~~~~~~~~~~~~~~~~~~~~

  In [1]: import arviz as az
     ...: schools = az.load_arviz_data("centered_eight")
     ...: az.summary(schools)

> ArviZ supports label based indexing.

Reviewer: I would like to still be explicit about xarray. Many users only use ArviZ functions, plots and diagnostics and are not aware that all of that is powered by xarray under the hood. This is particularly relevant for researchers who then don't cite xarray when publishing even though they did cite ArviZ. It is not "our" problem per se, but I think we should try to nudge users towards the right citations and it is also bad for us indirectly because xarray is generally undercited which dificults it getting grants and slows down its development.

Through label based indexing you can use labels to plot a subset of selected variables.

Example: Label based indexing
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

For a case where the coordinate values shown for the theta variable correspond to the `school` dimension
you can plot only the ``tau`` variable to better inspect its 1.03 :func:`~arviz.rhat` value,

> only the ``theta`` variable to the ``Choate`` and ``St. Paul's`` coordinates, which have higher means, you can ploy only the ``theta`` variable by running the following command:

Reviewer: My original description wasn't very good, but it tried to say that we are plotting all the things that were described, we are not plotting only the `theta` variable. How about something like:

```
Given that summary by default shows the coordinate values and not their indices or some aliases, 
you can use what you see to subset the data. 
The summary shows that ``tau`` has 1.03 :func:`~arviz.rhat` value which deserves closer inspection. 
You can indicate ArviZ to plot ``tau`` by including it in the ``var_names`` argument. 
Moreover, you may also want to inspect closely the ``theta`` values for the ``Choate`` 
and ``St. Paul's`` coordinates because of their higher means. 
You can do so by including ``theta`` too in ``var_names`` and 
using the ``coords`` argument to select only these two coordinate values for the ``school`` dimension. 
You can generate this plot with the following command:
```

.. ipython:: python

    @savefig label_guide_plot_trace.png
    az.plot_trace(schools, var_names=["tau", "theta"], coords={"school": ["Choate", "St. Paul's"]}, compact=False);

> With this you can now identify issues for the low ``tau`` values.

Reviewer: ```suggestion
With this you can now identify issues for low ``tau`` values.
```

Example: Using the labeller argument
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> To create a report on Deerfield, Hotchkiss and Lawrenceville schools for the probability of ``theta > 5``, use the labeller argument to customize labels. 

Reviewer: ```suggestion
You'll now create a report on Deerfield, Hotchkiss and Lawrenceville schools for the probability of ``theta > 5`` and use the labeller argument to customize labels.
```

     ...: schools = az.load_arviz_data("centered_eight")
     ...: az.summary(schools)

ArviZ supports label based indexing powered by xarray.
Through label based indexing you can use labels to plot a subset of selected variables.

Example: Label based indexing
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

For a case where the coordinate values shown for the ``theta`` variable coordinate to the ``school`` dimension
you can indicate ArviZ to plot ``tau`` by including it in the ``var_names`` argument to inspect its 1.03 :func:`~arviz.rhat` value.
To inspect the ``theta`` values for the ``Choate`` and ``St. Paul's`` coordinates, you can include ``theta`` in ``var_names`` and use the ``coords`` argument to select only these two coordinate values.
You can generate this plot with the following command:

.. ipython:: python

    @savefig label_guide_plot_trace.png
    az.plot_trace(schools, var_names=["tau", "theta"], coords={"school": ["Choate", "St. Paul's"]}, compact=False);

With this you can now identify issues for low ``tau`` values.

Example: Using the labeller argument
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> To create a report on Deerfield, Hotchkiss and Lawrenceville schools for the probability of ``theta > 5`` and use the labeller argument to customize labels. 

Reviewer: This sentence now feels incomplete. It starts with "to create" and a description on what to create but nothing on how to create it.

A bit more context in case it helps. Here there are two goals: creating a report of theta > 5 and using the labeller to customize the labels. 

One can be done without the other, but instead of showing how to use the labeller (which I think should be done in the example in he docstring that is still wip), I wanted to show how to use the labeller to solve a more specific and real task. When exploring the mode ourselves, we won't generally care much about the labels, having the same thing as the code is fine. However, if generating a report to be published or presented, we'll probably want to take better care of the presentation, and match the labels to the labels in the equations of the paper instead of variables in the code. Hence the MapLabeller

Author: Thank you for providing a bit of context to this matter. The way I see it, since this piece of text is inside of the example, we don't even need the first part of the sentence.

    @savefig label_guide_plot_posterior.png
    In [1]: az.plot_posterior(schools, var_names="theta", coords=coords, labeller=labeller, ref_val=5);

Additional Resources
~~~~~~~~~~~~~~~~~~~~

- For a list of labellers available in ArviZ, see the :ref:`the API reference page <labeller_api>`.

> - For common customization classes, see the ``arviz.labels`` module.

Reviewer: The api reference has the docs for the `arviz.labels` module, the original last line tried to give a choice to readers. At this point they have seen how to use the `labeller` argument. Therefore, if they only want to do some more or less simple changes to the labels, they don't need to continue reading this guide, but they will probably need to read the api reference to find the class that suits their needs (which may be IdxLabeller instead of MapLabeller). Does this make sense? Maybe deleting this 2nd bullet point is enough already?

Also, how about using the `seealso` directive? https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-seealso

Author: Thank you for the hint with the `seealso` directive! This is what I was looking for.

Sorting labels
--------------

ArviZ allows labels to be sorted in two ways:

- Using the arguments passed to ArviZ plotting functions 

> - Sorting the underlying xarray Dataset

Reviewer: ```suggestion
- Sorting the underlying :class:`xarray.Dataset`
```

That was (yet another) oversight on my part when writing the guide.

The first option is more suitable for single time ordering whereas the second option is more suitable for sorting plots consistently.

.. note::

  Both ways are limited.

>   Multidimensional variables can not be separated from each other. 

Reviewer: ```suggestion
  Multidimensional variables can not be separated. 
```

Sorting coordinate values
.........................

> To sort coordinate values you have to define the order, store it, and use the result to sort the coordinate values. 

Reviewer: ```suggestion
To sort coordinate values you have to define the order, store it, and use the result to sort the coordinate values.
The order can be defined by performing some operations on our xarray objects (like it is shown in the example below)
or by manually creating a list with the desired order.
```

Example: Sorting the schools by mean
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

1) Locate the means of each school by using the following command:

.. ipython::

    In [1]: school_means = schools.posterior["theta"].mean(("chain", "draw"))
       ...: school_means

2) You can use the DataArray result to sort the coordinate values for ``theta``.
There are two ways of sorting:

.. tabbed:: ArviZ args

  Sort the coordinate values to pass them as a `coords` argument and choose the order of the rows.

>   To manually sort the schools, `sorted_schools`, define sorted_schools as a list.

Reviewer: To keep the example only about sorting by mean, I propose to move this above (I am commenting later in time on the above section in hopes that github will preserve the chronology).

```suggestion
```

Sorting coordinate values
.........................

> To sort coordinate values by mean you have to locate the means of each school and then use the DataArray result to sort the coordinate values. 

Reviewer: ```suggestion
To sort coordinate values you have to first define the order and store it, then use that result to sort the coordinate values. 
```

Example: Sorting the schools by mean
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

1) Locate the means of each school by using the following command:

.. ipython::

    In [1]: school_means = schools.posterior["theta"].mean(("chain", "draw"))
       ...: school_means

2) You can use the DataArray result to sort the coordinate values for ``theta``.
There are two ways of sorting:

.. tabbed:: ArviZ args

> Sort the coordinate values to pass them as a `coords` argument and choose the order of the rows.
> To manually sort the schools, `sorted_schools`, define sorted_schools as a list.

Reviewer: ```suggestion
  Sort the coordinate values to pass them as a `coords` argument and choose the order of the rows.
  To manually sort the schools, `sorted_schools`, define sorted_schools as a list.
```

This needs to be indented for rST to parse the tabs correctly

.. tabbed:: xarray

> You can use the :meth:`~xarray.Dataset.sortby` method to order our coordinate values directly at the source.

Reviewer: ```suggestion
  You can use the :meth:`~xarray.Dataset.sortby` method to order our coordinate values directly at the source.
```
